File: Python/src/Student.py
import logging
import random
import CompulsoryCourse
import Transcript
import Schedule
import Advisor
from typing import List

class Student:
    def __init__(self, student_name: str, student_surname: str, transcript_before: 'Transcript', transcript_after: 'Transcript',
                 schedule: 'Schedule', advisor: 'Advisor', enrolled_year: str) -> None:
        self.__student_name = student_name
        self.__student_surname = student_surname
        self.__transcript_before = transcript_before
        self.__transcript_after = transcript_after
        self.__schedule = schedule
        self.__advisor = advisor
        self.__enrolled_year = enrolled_year

    def __init__(self, student_name: str, student_surname: str, transcript_before: 'Transcript',
                 schedule: 'Schedule', advisor: 'Advisor', enrolled_year: str) -> None:
        self.__student_name = student_name
        self.__student_surname = student_surname
        self.__transcript_before = transcript_before
        self.__schedule = schedule
        self.__advisor = advisor
        self.__enrolled_year = enrolled_year

    def __init__(self, student_id: str, student_name: str, student_surname: str, enrolled_year: str) -> None:
        self.__student_name = student_name
        self.__student_surname = student_surname
        self.__enrolled_year = enrolled_year

    def get_studentID(self):
        return self.__studentID

    def set_studentID(self, studentID):
        self.__studentID = studentID

    def get_studentSurname(self):
        return self.__studentSurname

    def set_studentSurname(self, studentSurname):
        self.__studentSurname = studentSurname

    def get_studentName(self):
        return self.__studentName

    def set_studentName(self, studentName):
        self.__studentName = studentName

    def get_transcriptBefore(self):
        return self.__transcriptBefore

    def set_transcriptBefore(self, transcriptBefore):
        self.__transcriptBefore = transcriptBefore

    def get_transcriptAfter(self):
        return self.__transcriptAfter

    def set_transcriptAfter(self, transcriptAfter):
        self.__transcriptAfter = transcriptAfter

    def get_schedule(self):
        return self.__schedule

    def set_schedule(self, schedule):
        self.__schedule = schedule

    def get_advisor(self):
        return self.__advisor

    def set_advisor(self, advisor):
        self.__advisor = advisor

    def get_enrolledYear(self):
        return self.__enrolledYear

    def set_enrolledYear(self, enrolledYear):
        self.__enrolledYear = enrolledYear

    def create_studentID(self):
        department = "1501"
        self.set_studentID(department + self.__enrolledYear[2:] + str(int(random.random() * 899) + 100))

    def calculateTranscriptAfter(self, enrollmentRequest, prob):
        student = enrollmentRequest.getStudent()
        transcriptBefore = student.getTranscriptBefore()
        logger = logging.getLogger(__name__)

        temp = Transcript()

        courses = enrollmentRequest.getCourses()
        taken_credit = 0
        for credit in courses.map(CompulsoryCourse.getCredit):
            taken_credit += credit
        logger.info(f"End of the term: Student {student.getStudentID()} passed the courses: {courses.map(CompulsoryCourse.getCourseCode)}")
        temp.setTakenCredit(transcriptBefore.getTakenCredit() + taken_credit)
        temp.addCourseGrade(temp.randomCourseGrade(courses, prob))
        temp.addCourseGrade(transcriptBefore.getCourseGrades())
        temp.addTakenCourse(courses)
        temp.addTakenCourse(transcriptBefore.getTakenCourses())
        temp.setGpa(temp.calculateGpa())
        student.setTranscriptAfter(temp)
        logger.info("New GPA is calculated and transcript is updated.")

    def __str__(self):
        return "Student{studentID='{}', studentName='{}', studentSurname='{}', transcriptBefore={}, transcriptAfter={}, schedule={}, advisor={}, enrolledYear='{}'}".format(self.__studentID, self.__studentName, self.__studentSurname, self.__transcriptBefore, self.__transcriptAfter, self.__schedule, self.__advisor, self.__enrolledYear)

File: Java/Java Project/Iteration_2/src/Student.java
package Iteration_2.src;


import org.apache.log4j.Logger;

import java.util.ArrayList;

public class Student {

    private String studentID;
    private String studentName;
    private String studentSurname;
    private Transcript transcriptBefore;
    private Transcript transcriptAfter;
    private Schedule schedule;
    private Advisor advisor;
    private String enrolledYear;

    public Student(String studentName, String studentSurname, Transcript transcriptBefore, Transcript transcriptAfter,
                   Schedule schedule, Advisor advisor, String enrolledYear) {


        this.studentName = studentName;
        this.studentSurname = studentSurname;
        this.transcriptBefore = transcriptBefore;
        this.transcriptAfter = transcriptAfter;
        this.schedule = schedule;
        this.advisor = advisor;
        this.enrolledYear = enrolledYear;
    }
    public Student(String studentName, String studentSurname, Transcript transcriptBefore,
                   Schedule schedule, Advisor advisor, String enrolledYear) {


        this.studentName = studentName;
        this.studentSurname = studentSurname;
        this.transcriptBefore = transcriptBefore;
        this.schedule = schedule;
        this.advisor = advisor;
        this.enrolledYear = enrolledYear;
    }

    public Student(String studentID, String studentName, String studentSurname, String enrolledYear) {

        this.studentName = studentName;
        this.studentSurname = studentSurname;
        this.enrolledYear = enrolledYear;
    }

    public String getStudentID() {
        return this.studentID;
    }

    public void setStudentID(String studentID) {
        this.studentID = studentID;
    }

    public String getStudentSurname() {
        return this.studentSurname;
    }

    public void setStudentSurname(String studentSurname) {
        this.studentSurname = studentSurname;
    }

    public String getStudentName() {
        return this.studentName;
    }

    public void setStudentName(String studentName) {
        this.studentName = studentName;
    }

    public Transcript getTranscriptBefore() {
        return transcriptBefore;
    }

    public void setTranscriptBefore(Transcript transcriptBefore) {
        this.transcriptBefore = transcriptBefore;
    }

    public Transcript getTranscriptAfter() {
        return transcriptAfter;
    }

    public void setTranscriptAfter(Transcript transcriptAfter) {
        this.transcriptAfter = transcriptAfter;
    }

    public Schedule getSchedule() {
        return this.schedule;
    }

    public void setSchedule(Schedule schedule) {
        this.schedule = schedule;
    }

    public Advisor getAdvisor() {
        return this.advisor;
    }

    public void setAdvisor(Advisor advisor) {
        this.advisor = advisor;
    }

    public String getEnrolledYear() {
        return this.enrolledYear;
    }

    public void setEnrolledYear(String enrolledYear) {
        this.enrolledYear = enrolledYear;
    }

    public Student() {
    }

    public void createStudentID() {
        String department = "1501";
        this.setStudentID(department + this.enrolledYear.substring(2) + String.valueOf((int)(Math.random() * 899)+100));

    }
    public void calculateTranscriptAfter(EnrollmentRequest enrollmentRequest,double prob) {
        Student student = enrollmentRequest.getStudent();
        Transcript transcriptBefore = student.getTranscriptBefore();
        Logger logger= Logger.getLogger(Student.class.getName());

        Transcript temp = new Transcript();

        ArrayList<CompulsoryCourse> courses = enrollmentRequest.getCourses();
        int takenCredit = 0;
        for(int credit : courses.stream().map(CompulsoryCourse::getCredit).toList())
            takenCredit += credit;
        logger.info("End of the term: Student "+ student.getStudentID()+" passed the courses: "+ courses.stream().map(CompulsoryCourse::getCourseCode).toList());
        temp.setTakenCredit(transcriptBefore.getTakenCredit()+takenCredit);
        temp.addCourseGrade(temp.randomCourseGrade(courses,prob));
        temp.addCourseGrade(transcriptBefore.getCourseGrades());
        temp.addTakenCourse(courses);
        temp.addTakenCourse(transcriptBefore.getTakenCourses());
        temp.setGpa(temp.calculateGpa());
        student.setTranscriptAfter(temp);
        logger.info("New GPA is calculated and transcript is updated. ");



    }

    @Override
    public String toString() {
        return "Student{" +
                "studentID='" + studentID + '\'' +
                ", studentName='" + studentName + '\'' +
                ", studentSurname='" + studentSurname + '\'' +
                ", transcriptBefore=" + transcriptBefore +
                ", transcriptAfter=" + transcriptAfter +
                ", schedule=" + schedule +
                ", advisor=" + advisor +
                ", enrolledYear='" + enrolledYear + '\'' +
                '}';
    }

}
